sentinel-core/src/main/java/com/alibaba/csp/sentinel/slots/block/degrade/circuitbreaker/AbstractCircuitBreaker.java

```diff
@@ -99,11 +103,24 @@ protected boolean fromCloseToOpen(double snapshotValue) {
         return false;
     }
 
-    protected boolean fromOpenToHalfOpen() {
+    protected boolean fromOpenToHalfOpen(Context context) {
         if (currentState.compareAndSet(State.OPEN, State.HALF_OPEN)) {
             for (CircuitBreakerStateChangeObserver observer : observerRegistry.getStateChangeObservers()) {
                 observer.onStateChange(State.OPEN, State.HALF_OPEN, rule, null);
             }
+            Entry entry = context.getCurEntry();
+            entry.whenComplete(new BiConsumer<Context, Entry>() {
+                
+                @Override
+                public void accept(Context context, Entry entry) {
+                    if (entry.getBlockError() != null) {
+                        // blocked
+                        currentState.compareAndSet(State.HALF_OPEN, State.OPEN);
```

Comment:
Maybe the state changes need to notify **CircuitBreakerStateChangeObserver**.

Reply:
> Maybe the state changes need to notify **CircuitBreakerStateChangeObserver**.

Yes, absolutely.

It has been updated.

Reply:
It's required, but this could be confusing for users (as the internal reverting is not expected for users)😂

Reply:
> It's required, but this could be confusing for users (as the internal reverting is not expected for users)😂

Yeah indeed but the conditions would be hard to met.